fix!: align `target` and `baseElement` options with testing-library

mcous opened the pull request:
Fixes #312, fixes #313

BREAKING CHANGES: The `container` option has been renamed to `baseElement`,
  `result.container` is now set to `target` rather than `baseElement`,
  and `render` will now throw if you mix props with the `target` option.

## Change log

- Renamed `renderOptions.container` to `renderOptions.baseElement`
- Return `baseElement` as `result.baseElement`
- Default `baseElement` to `componentOptions.target` if used, else `document.body`
- Return `componentOptions.target` as `result.container`, which is one layer lower in the DOM than the previous behavior and matches other `testing-library` frameworks
- `componentOptions.target` will now trigger an error if it is mixed with props, matching the behavior of other component options
- Updated and reorganized tests
- Updated type definitions

mcous commented on package.json:
```diff
@@ -94,7 +94,7 @@
     "prettier": "3.2.4",
     "prettier-plugin-svelte": "3.1.2",
-    "svelte": "^4.2.10",
+    "svelte": "^3 || ^4 || ^5",
```
I inadvertently changed this in #321. Keeping the dev version wide makes it easier to switch between versions locally when testing

mcous commented on src/__tests__/__snapshots__/render.test.js.snap:
Snapshots are inconsistent between Svelte 4 and Svelte 5, and the tests themselves that used them were better served by actually querying the DOM structure, so I removed them. I don't think it's really our job to test that Svelte writes the the DOM properly, which was the bulk of the snapshot

yanick replied:
:+1: If we ever get in the situation where we want to have the snapshots for different svelte versions, I guess we could always do

```
expect(blah).toMatchSnapshot( `for Svelte ${SVELTE_VERSION}` );
```

mcous replied:
Oh, I use snapshots so rarely that I didn't even know that was an option 😅. Can bring them back if you think they're useful!

yanick replied:
Nah. I agree with you that the exact shape of the DOM is not this library's worry. 

I was thinking of the general case and what we can do if we do want to have one of those tests work for several svelte versions. Access to the version of all packages is one thing I wish the js ecosystem did better.

mcous replied:
We could probably do something clever with workspaces to be able to install different Svelte versions locally at the same time and run the same test suite against them. It'd be non-trivial setup though, and would likely start placing some stricter requirements on which package manager we use for dev. Something to approach with caution, if at all

mcous commented on src/__tests__/fixtures/Comp.svelte:
```diff
@@ -1,23 +1,17 @@
 <svelte:options accessors />
 
 <script>
-  import { getContext } from 'svelte'
```
Duplicate context testing between `render.test.js` and `context.test.js`. Since `context.test.js` has its own `Context.svelte` fixture, I removed context stuff from here

mcous commented on src/__tests__/fixtures/Comp2.svelte:
This fixture wasn't being used for anything meaningful that I could figure out. I suspect it was from an old CJS vs ESM test. See other comment

mcous commented on src/__tests__/fixtures/Rerender.svelte:
Fixture was no longer needed now that `rerender` does not unmount. Instead, we can just hold on to the DOM element and make sure the element itself is updated rather than replaced

mcous commented on src/__tests__/render.test.js:
```diff
@@ -1,123 +1,84 @@
+    expect(baseElement.firstChild).toBe(container)
   })
 
-  test('correctly find component constructor on the default property', () => {
-    const { getByText } = render(CompDefault, { props: { name: 'World' } })
```
This is the test that was using `Comp2.svelte`, but `Comp.svelte` and `Comp2.svelte` had the same contents and were imported in the same manner

```js
import Comp from './fixtures/Comp.svelte'
import CompDefault from './fixtures/Comp2.svelte'
```

Since I couldn't figure out what the test was doing, and I couldn't figure out a way to replicate a CJS-style ESM export without causing some very weird errors, I dropped the test

yanick replied:
I think what happened there is that the original test was importing from `comp.html`, and esm wouldn't do that so I had to use a `.svelte` file instead?

But yeah, that test can definitively go.

mcous commented on src/pure.js:
```diff
@@ -11,45 +11,21 @@ const componentCache = new Set()
 
-    return { props: options }
-  }
+  const ComponentConstructor = Component.default || Component
```
I think this must be a weird holdover from CJS days. I may remove it in a future PR, since I can't figure out how to actually trigger this code-path in tests now that we're ESM-native

mcous commented on src/__tests__/rerender.test.js:
```diff
@@ -1,41 +1,50 @@
-  rerender({ props: { name: 'World 2' } })
-  await expectToRender('Hello World 2!')
-  expect(onDestroyed).not.toHaveBeenCalled()
+  test('change props with accessors', async () => {
```
This test moved over from `render.test.js` because it has more in common with `rerender.test.js`

mcous commented:
@yanick let me know if you think this is too much to change in one PR; I can split it up into a few separate changes relatively trivially

yanick reviewed (approved):
Looks good! Can you also prepare a branch for the documentation changes related to this PR?

mcous commented:
10-4, will undraft this PR when corresponding docs PR is up

mcous commented:
Docs PR up at testing-library/testing-library-docs#1369

mcous commented:
Giving this one last smoke test in a real-life suite, will update here with results

mcous commented:
Passed our ~1000 test suite without any unexpected failures nor type-checking issues (still on Svelte v4). Had exactly one failure: a test that was using `rerender` and relying on the old synchronous destroy and remount behavior. This, in turn, uncovered an issue with the type definitions for `rerender` - in reality, it accepts `Partial<Props>`, not `Props`.

Switching the the new `await rerender(...)` fixed the test (and got the linter to stop complaining), and I just pushed the type definitions fix, so I think we're all good here if CI agrees with me

yanick commented:
@mcous I tried to merge your branch with the head of next on #331  I think I got it all?

mcous commented:
@yanick thanks for the assist. I squashed and rebased this branch while referencing #331, which was a little easier on my brain. It looked like 331 added has a few duplicate test cases, but otherwise I mostly followed the changes there

mcous commented on src/pure.js:
```diff
@@ -80,27 +86,23 @@ export class SvelteTestingLibrary {
-        this.componentCache.delete(component)
-      })
-    }
+    component.$$.on_destroy.push(() => {
+      this.componentCache.delete(component)
+    })
```
Should I just remove this now? It's a weird little bit of logic that makes it so one can call _either_ `component.$destroy` or `unmount` in a test. It feels reasonable to only provide one way - `unmount` - to do this

yanick replied:
I'd tend to say "nah". It's been the established behavior, and with Svelte 4 probably being eclipsed by Svelte 5 soon, there is little gain to change the behavior now.

mcous commented:
@yanick heads up, looks like [semantic release missed this](https://github.com/testing-library/svelte-testing-library/actions/runs/8316270863/job/22755541762#step:5:55):

```
[2:54:51 PM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analyzing commit: fix!: align `target` and `baseElement` options with testing-library (#325)

Fixes #312, fixes #313

BREAKING CHANGES: The `container` option has been renamed to `baseElement`,
  `result.container` is now set to `target` rather than `baseElement`,
  and `render` will now throw if you mix props with the `target` option.
[2:54:51 PM] [semantic-release] [@semantic-release/commit-analyzer] › ℹ  The commit should not trigger a release
```

Suspected causes:

- semantic release [does not recognize the `!`](https://github.com/semantic-release/semantic-release/issues/2339) as a breaking change in `fix!: ...`, so it didn't see `fix` nor that it was a major bump
- I accidentally wrote `BREAKING CHANGES:` instead of `BREAKING CHANGE:` in the footer, so it also missed the major bump here

Not sure the best way to fix this up, the quickest / easiest might be a little distasteful history rewrite on `next`
